File: cli/src/core/references/sources/GitHubNormalize.test.ts
import { describe, expect, it } from "vitest";
import { reshapeGitHubIssue } from "./GitHubNormalize.js";

const reshape = (raw: unknown) => reshapeGitHubIssue(raw) as Record<string, unknown>;

describe("reshapeGitHubIssue", () => {
	it("reshapes a single `_fetch_issue` payload (unwrap issue.*, rename, flatten)", () => {
		const out = reshape({
			issue: {
				issue_number: 12,
				title: "T",
				url: "https://github.com/o/r/issues/12",
				body: "B",
				state: "open",
				// bare string kept, empty dropped, number dropped, {name} flattened, {nope} dropped
				labels: ["bug", "", 7, { name: "feat" }, { nope: 1 }],
				assignees: [{ login: "alice" }],
				repository_full_name: "o/r",
			},
		});
		expect(out.number).toBe(12);
		expect(out.title).toBe("T");
		expect(out.html_url).toBe("https://github.com/o/r/issues/12");
		expect(out.body).toBe("B");
		expect(out.state).toBe("open");
		expect(out.labels).toEqual(["bug", "feat"]);
		expect(out.assignees).toEqual(["alice"]);
		expect(out.repository).toEqual({ full_name: "o/r" });
	});

	it("reads a flat issue (no `issue` wrapper) and `repository_full_name` from the raw object", () => {
		const out = reshape({
			number: 5,
			title: "Flat",
			html_url: "https://github.com/o/r/issues/5",
			repository_full_name: "o/r",
		});
		expect(out.number).toBe(5);
		expect(out.html_url).toBe("https://github.com/o/r/issues/5");
		expect(out.repository).toEqual({ full_name: "o/r" });
	});

	it("derives `number` from an issue URL when it is missing (search hit)", () => {
		const out = reshape({ url: "https://github.com/o/r/issues/959", number: null, title: "S" });
		expect(out.number).toBe(959);
		expect(out.html_url).toBe("https://github.com/o/r/issues/959");
	});

	it("derives `number` from a pull-request URL too", () => {
		const out = reshape({ url: "https://github.com/o/r/pull/77", title: "PR" });
		expect(out.number).toBe(77);
	});

	it("keeps an explicit `number` and does not overwrite from URL", () => {
		const out = reshape({ number: 3, html_url: "https://github.com/o/r/issues/959", title: "x" });
		expect(out.number).toBe(3);
	});

	it("leaves `number` undefined when there is no URL or the URL has no issue/PR number", () => {
		expect(reshape({ title: "no url" }).number).toBeUndefined();
		expect(reshape({ title: "blob", url: "https://github.com/o/r/blob/main/x.ts" }).number).toBeUndefined();
	});

	it("omits labels/assignees when no usable entries survive", () => {
		const out = reshape({
			number: 1,
			title: "x",
			url: "https://github.com/o/r/issues/1",
			labels: ["", 7, { nope: 1 }],
		});
		expect(out.labels).toBeUndefined();
		expect(out.assignees).toBeUndefined();
	});

	it("returns non-object input unchanged", () => {
		expect(reshapeGitHubIssue(123)).toBe(123);
		expect(reshapeGitHubIssue(null)).toBe(null);
		expect(reshapeGitHubIssue("x")).toBe("x");
	});
});

File: intellij/src/test/kotlin/ai/jolli/jollimemory/core/references/GitHubNormalizeTest.kt
package ai.jolli.jollimemory.core.references

import com.google.gson.JsonParser
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test

class GitHubNormalizeTest {

	private fun parse(json: String) = JsonParser.parseString(json)

	@Nested
	inner class Unwrap {
		@Test
		fun `unwraps Codex double-wrapped issue`() {
			val raw = parse("""{"issue":{"issue_number":42,"title":"Bug","url":"https://github.com/o/r/issues/42","body":"desc","state":"open","labels":["bug"],"assignees":["alice"],"repository_full_name":"o/r"}}""")
			val result = GitHubNormalize.reshape(raw)!!.asJsonObject
			result.intOrNull("number") shouldBe 42
			result.stringOrNull("title") shouldBe "Bug"
			result.stringOrNull("html_url") shouldBe "https://github.com/o/r/issues/42"
			result.stringOrNull("body") shouldBe "desc"
			result.stringOrNull("state") shouldBe "open"
			result.objectOrNull("repository")!!.stringOrNull("full_name") shouldBe "o/r"
		}

		@Test
		fun `flat form passes through with renames`() {
			val raw = parse("""{"number":7,"title":"Fix","html_url":"https://github.com/o/r/issues/7","state":"closed"}""")
			val result = GitHubNormalize.reshape(raw)!!.asJsonObject
			result.intOrNull("number") shouldBe 7
			result.stringOrNull("html_url") shouldBe "https://github.com/o/r/issues/7"
		}
	}

	@Nested
	inner class LabelFlattening {
		@Test
		fun `flattens object labels`() {
			val raw = parse("""{"number":1,"title":"T","html_url":"https://github.com/o/r/issues/1","labels":[{"name":"bug"},{"name":"p1"}]}""")
			val result = GitHubNormalize.reshape(raw)!!.asJsonObject
			val labels = result.arrayOrNull("labels")!!
			labels.size() shouldBe 2
			labels[0].asString shouldBe "bug"
		}

		@Test
		fun `flattens mixed labels, filters empty`() {
			val raw = parse("""{"number":1,"title":"T","html_url":"https://github.com/o/r/issues/1","labels":["ok",{"name":""},{"name":"good"}]}""")
			val result = GitHubNormalize.reshape(raw)!!.asJsonObject
			val labels = result.arrayOrNull("labels")!!
			labels.size() shouldBe 2
		}
	}

	@Nested
	inner class NumberDerivation {
		@Test
		fun `derives number from URL when missing`() {
			val raw = parse("""{"title":"T","url":"https://github.com/o/r/issues/99"}""")
			val result = GitHubNormalize.reshape(raw)!!.asJsonObject
			result.intOrNull("number") shouldBe 99
		}

		@Test
		fun `derives number from pull URL`() {
			val raw = parse("""{"title":"T","url":"https://github.com/o/r/pull/55"}""")
			val result = GitHubNormalize.reshape(raw)!!.asJsonObject
			result.intOrNull("number") shouldBe 55
		}

		@Test
		fun `does not override explicit number`() {
			val raw = parse("""{"number":10,"title":"T","url":"https://github.com/o/r/issues/99"}""")
			val result = GitHubNormalize.reshape(raw)!!.asJsonObject
			result.intOrNull("number") shouldBe 10
		}
	}

	@Test
	fun `null input returns null`() {
		GitHubNormalize.reshape(null) shouldBe null
	}

	@Test
	fun `non-object passes through`() {
		val arr = parse("[1,2]")
		GitHubNormalize.reshape(arr) shouldBe arr
	}
}
